library/std/src/process.rs

```diff
@@ -1465,6 +1488,117 @@ impl fmt::Display for ExitStatus {
     }
 }
 
+/// Allows extension traits within `std`.
+#[unstable(feature = "sealed", issue = "none")]
+impl crate::sealed::Sealed for ExitStatusError {}
+
+/// Describes the result of a process after it has failed
+///
+/// Produced by the [`.exit_ok`](ExitStatus::exit_ok) method on [`ExitStatus`].
+///
+/// # Examples
+///
+/// ```
+/// #![feature(exit_status_error)]
+/// # if cfg!(unix) {
+/// use std::process::{Command, ExitStatusError};
+///
+/// fn run(cmd: &str) -> Result<(),ExitStatusError> {
+///     Command::new(cmd).status().unwrap().exit_ok()?;
+///     Ok(())
+/// }
+///
+/// run("true").unwrap();
+/// run("false").unwrap_err();
+/// # } // cfg!(unix)
+/// ```
+#[derive(PartialEq, Eq, Clone, Copy, Debug)]
+#[unstable(feature = "exit_status_error", issue = "none")]
+// The definition of imp::ExitStatusError should ideally be such that
+// Result<(), imp::ExitStatusError> has an identical representation to imp::ExitStatus.
+pub struct ExitStatusError(imp::ExitStatusError);
```

Comment:
I feel like we might want to just call this a `NonZeroExitStatus` and still implement `Error` for it, though I'm conflicted because I know if we have any other types that implement `Error` but don't have `Error` in their type name.

Reply:
> I feel like we might want to just call this a `NonZeroExitStatus` and still implement `Error` for it, though I'm conflicted because I don't know if we have any other types that implement `Error` but don't have `Error` in their type name.

I'm pretty sure calling it `ExitStatusError` is better.  We have lots and lots of things called `SuchAndSuchError`, many of which are units or newtypes around something else.  And the name `NonZeroExitStatus` tells us rather more about the underlying representation than ideal.  Or, I think, even, portable.  In theory there might be some platform where the underlying raw status for a success is 0xffffffff or something, or where the process termination information isn't even a plain number.